clients/da-vinci-client/src/main/java/com/linkedin/davinci/kafka/consumer/LeaderFollowerStoreIngestionTask.java

```diff
@@ -2066,6 +2070,36 @@ private void propagateHeartbeatFromUpstreamTopicToLocalVersionTopic(
     }
   }
 
+  protected void recordHeartbeatReceived(
+      PartitionConsumptionState partitionConsumptionState,
+      PubSubMessage<KafkaKey, KafkaMessageEnvelope, Long> consumerRecord,
+      String kafkaUrl) {
+    if (heartbeatMonitoringService == null) {
+      // Not enabled!
+      return;
+    }
+    long producerTimeStamp =
+        max(consumerRecord.getPubSubMessageTime(), consumerRecord.getValue().producerMetadata.messageTimestamp);
```

Comment:
Related to the comment to potentially move `recordHeartbeatReceived` to when we receive the HB in drainer. In that case maybe the `consumerRecord.getValue().producerMetadata.messageTimestamp` would get messed up since it will be the timestamp when the record was copied into VT from RT?

Reply:
I guess it would depend on what we queue to drainer.  We could record it into what we queue there maybe.

Reply:
I'm not following here.  Isn't `producerMetadata.messageTimestamp` always older/smaller than `getPubSubMessageTime()`? What's the circumstances that a `producerMetadata.messageTimestamp` being unreliable?

Reply:
I actually followed this pattern from another bit of code which was doing this (I assumed the original author thought of some edge case).  The comment above it said something about clock skew being a potential concern, though I don't think using Max would completely alleviate that notion besides erring on the side of caution.

Reply:
Added `max()` to alleviate the time skew occurring in the leader topic partition, which would causing followers to consider the heartbeat as stale (since they check for one minute freshness) and never become ready to serve. This is not a complete fix, but the chance of both the leader topic partition and Kafka broker's time going awry at the same time is rare or if it happens, there will likely be other major issues around it.
For monitoring purposes, having either the producer timestamp, broker timestamp, or the maximum of both doesn't really make a difference in my opinion, unless we need to monitor this delay by calculating it from two separate metrics. Therefore, this change looks good IMO. 

Nit: For clarity and to keep the reasoning together, we can convert this line into a small function and move the comments about the skew there.